planning/MartenV4-WastazComments.md

```diff
@@ -0,0 +1,174 @@
+The number 1 realization that I _hope_ more of the .NET community is finally going to realize as they start using this feature more is that _reading data from a database is external input_.
+
+This leads to basically two choices. Either you can make your document/event model the optimistic or the pessmistic way:
+
+```csharp
+public class NutrientInfo { /* class omitted for brevity */ }
+
+// Optimistic
+public class FruitDocument {
+  public Guid Id { get; set; }
+  public string Name { get; set; }
+  public int Ranking { get; set; }
+  public NutrientInfo { get; set; }
+}
+
+// Pessimistic
+public class FruitDocument {
+  public Guid Id { get; set; }
+  public string? Name { get; set; }
+  public int Ranking { get; set; }
+  public NutrientInfo? { get; set; }
+}
+```
+
+So which should you do? Honestly Id say the pessimistic one, because we are deserializing json here. And even in the optimistic one you _probably_ check the value of `Ranking` depending on
+what values are valid in your domain because thats gonna be `default(int)` if its missing. Now, _this has always been a problem_ but NRTs really shove it in your face when you realize that
+you probably need to mark a ton of shit as nullable in every document/event you have.
+
+We have tried both using optimistic and pessimistic models. Pessimistic models leads to...code that you dont really want. But optimistic models however leads to either a false sense of
+security or you have to force a validation step for your document after each load. Thing that we think works best so far has been building validators with FluentValidation for every event/document
+type, using optimistic models and validating the documents after reading them. Its...not that fun, but it works and it lets us actually model the documents/events according to how they should look
+according to the domain.
+
+However there are some ways to make this easier.
+Ive seen the discussions on the "streaming api", but I think that what may be missing from these discussions is that there can actually be tremendous alternate value in providing a way to
+easily read raw documents/events as `string` or `Stream`. Because it allows an easier way of hooking in a _parser_, not a _serializer_. There's some good blog posts out there on the virtues of
+parsing instead of validating (or serializing) so Im not gonna repeat the points that they already made. But with this type of streaming api you could much easier hook up a parser which would not
+only help in having an easier way of constructing correct objects that respects NRTs, but could also help a lot with the F# side of this where immutability and discriminated unions are not always the
+best at serializing/deserializing. You can probably do some things like that with the `ISerializer` interface today, but having a way to just bypass serialization completely is quite powerful.
+
+# Event versioning
```

Comment:
As I wrote in another comment - we should support upcasters and downcasters out of the box for that goal. Forcing our users to do a JSON magic is far from ideal (especially that other ES systems supports that).

I was also thinking about some JSON Schema repository that'd allow to perform validations if the new event structure is compatible with the previous (to eg. catch some incompatibility).